// react-sizes/lib/SizesProvider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _contextTypes = require('./contextTypes');

var _contextTypes2 = _interopRequireDefault(_contextTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SizesProvider = function (_Component) {
  _inherits(SizesProvider, _Component);

  function SizesProvider() {
    _classCallCheck(this, SizesProvider);

    return _possibleConstructorReturn(this, (SizesProvider.__proto__ || Object.getPrototypeOf(SizesProvider)).apply(this, arguments));
  }

  _createClass(SizesProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      var _props$config = this.props.config,
          config = _props$config === undefined ? {} : _props$config;


      return _defineProperty({}, _contextTypes.contextKey, {
        fallbackWidth: config.fallbackWidth || null,
        fallbackHeight: config.fallbackHeight || null,
        throttle: config.throttle || 200
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return this.props.children;
    }
  }]);

  return SizesProvider;
}(_react.Component);

SizesProvider.childContextTypes = _contextTypes2.default;
exports.default = SizesProvider;
