// react-imported-component/dist/cjs/scanners/shared.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFileContent = exports.pWriteFile = exports.pReadFile = exports.getMatchString = exports.getRelative = exports.normalizePath = exports.promisify = void 0;

var _path = require("path");

var _fs = require("fs");

try {
  require('babel-polyfill');
} catch (err) {
  try {
    require('@babel/polyfill');
  } catch (e) {
    throw Error('react-imported-component: scanImports is requiring babel-polyfill, or @babel/polyfill to work. Please add this dependency.');
  }
}
/* eslint-disable no-console */


var promisify = function promisify(fn, context, noReject) {
  return function () {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return new Promise(function (resolve, reject) {
      fn.call.apply(fn, [context].concat(args, [function (error, ok) {
        if (error) {
          if (noReject) {
            resolve(error);
          } else {
            reject(error);
          }
        }

        resolve(ok);
      }]));
    });
  };
};

exports.promisify = promisify;

var normalizePath = function normalizePath(path) {
  return path.split(_path.sep).join('/');
};

exports.normalizePath = normalizePath;

var getRelative = function getRelative(from, to) {
  // force one unit paths
  var rel = normalizePath((0, _path.relative)(from, to));
  return rel[0] !== '.' ? './' + rel : rel;
};

exports.getRelative = getRelative;

var getMatchString = function getMatchString(pattern, selected) {
  return function (str) {
    return (str.match(new RegExp(pattern, 'g')) || []).map(function (statement) {
      return statement.match(new RegExp(pattern, 'i'))[selected];
    });
  };
};

exports.getMatchString = getMatchString;
var pReadFile = promisify(_fs.readFile);
exports.pReadFile = pReadFile;
var pWriteFile = promisify(_fs.writeFile);
exports.pWriteFile = pWriteFile;

var getFileContent = function getFileContent(file) {
  return pReadFile(file, 'utf8');
};

exports.getFileContent = getFileContent;
